client/service-container-chain/src/data_preservers.rs

```diff
@@ -0,0 +1,360 @@
+
+use {
+    crate::spawner::{ContainerChainSpawner, Spawner, TSelectSyncMode},
+    dc_orchestrator_chain_interface::{
+        DataPreserverAssignment, OrchestratorChainInterface, OrchestratorChainResult,
+    },
+    futures::stream::StreamExt,
+    std::future::Future,
+    tc_consensus::ParaId,
+};
+
+pub type ProfileId = <dancebox_runtime::Runtime as pallet_data_preservers::Config>::ProfileId;
+
+async fn try_fut<T, E>(fut: impl Future<Output = Result<T, E>>) -> Result<T, E> {
+    fut.await
+}
+
+/// Watch assignements by indefinitly listening to finalized block notifications and switching to
+/// the chain the profile is assigned to.
+pub async fn task_watch_assignment(spawner: impl Spawner, profile_id: ProfileId) {
+    use dc_orchestrator_chain_interface::DataPreserverAssignment as Assignment;
+
+    if let OrchestratorChainResult::Err(e) = try_fut(async move {
+        let orchestrator_chain_interface = spawner.orchestrator_chain_interface();
+
+        let mut current_assignment = DataPreserverAssignment::<ParaId>::NotAssigned;
+
+        let mut stream = orchestrator_chain_interface
+            .finality_notification_stream()
+            .await?;
+
+        while let Some(header) = stream.next().await {
+            let hash = header.hash();
+
+            let new_assignment = orchestrator_chain_interface
+                .data_preserver_active_assignment(hash, profile_id)
+                .await?;
+
+            log::info!("Assignement for block {hash}: {new_assignment:?}");
+
+            match (current_assignment, new_assignment) {
+                // no change
+                (x, y) if x == y => continue,
+                // switch from not assigned/inactive to active, start embeded node
+                (
+                    Assignment::NotAssigned | Assignment::Inactive(_),
+                    Assignment::Active(para_id),
+                ) => {
+                    spawner.spawn(para_id, false).await;
+                }
+                // Assignement switches from active to inactive for same para_id, we stop the
+                // embeded node but keep db
+                (Assignment::Active(para_id), Assignment::Inactive(x)) if para_id == x => {
+                    spawner.stop(para_id, true); // keep db
+                }
+                // No longer assigned or assigned inactive to other para id, remove previous node
+                (
+                    Assignment::Active(para_id),
+                    Assignment::Inactive(_) | Assignment::NotAssigned,
+                ) => {
+                    spawner.stop(para_id, false); // don't keep db
+                }
+                // Changed para id, remove previous node and start new one
+                (Assignment::Active(previous_para_id), Assignment::Active(para_id)) => {
+                    spawner.stop(previous_para_id, false); // don't keep db
+                    spawner.spawn(para_id, false).await;
```

Comment:
This case could cause issues with the ports being unavailable, because `stop` doesn't wait for the chain to stop. You can use `wait_for_paritydb_lock` to wait for the previous chain to stop.